Helix crashed unexpectedly when trying to <A-o> in Rust

### Summary

I tried to expand my selection to the surrounding tree-sitter node when helix crashed with this panic: 
```
thread 'main' panicked at helix-core/src/syntax/tree_cursor.rs:85:31:
invalid HopSlotMap key used
note: run with `RUST_BACKTRACE=1` environment variable to display a backtrace
```

### Reproduction Steps

~Unfortunately I wasn't able to reproduce the bug.~
**Update:** I was able to reproduce this, see below

I tried this:

1. `hx`
2. Edit some code
3. Try to expand selection in this line
```rust
PathBuf::from(format!(concat!(env!("SOLITAIRE_CARDS_LOCATION"), "/{value_char}{suit_char}.svg")))
```

I expected this to happen:
Selection gets expanded

Instead, this happened:
Helix crashed and immediately dumped me back into the shell.

### Helix log

<details><summary>~/.cache/helix/helix.log</summary>

```
2024-06-16T13:39:29.212 helix_term::handlers::signature_help [ERROR] signature help request failed: request 14 timed out
2024-06-16T13:39:31.974 helix_lsp::transport [ERROR] Tried sending response into a closed channel (id=Num(1)), original request likely timed out
2024-06-16T13:39:31.975 helix_lsp::transport [ERROR] Tried sending response into a closed channel (id=Num(5)), original request likely timed out
2024-06-16T13:39:31.975 helix_lsp::transport [ERROR] Tried sending response into a closed channel (id=Num(9)), original request likely timed out
2024-06-16T13:39:31.975 helix_lsp::transport [ERROR] Tried sending response into a closed channel (id=Num(8)), original request likely timed out
2024-06-16T13:39:31.975 helix_lsp::transport [ERROR] Tried sending response into a closed channel (id=Num(6)), original request likely timed out
2024-06-16T13:39:31.975 helix_lsp::transport [ERROR] Tried sending response into a closed channel (id=Num(3)), original request likely timed out
2024-06-16T13:39:31.975 helix_lsp::transport [ERROR] Tried sending response into a closed channel (id=Num(2)), original request likely timed out
2024-06-16T13:39:31.975 helix_lsp::transport [ERROR] Tried sending response into a closed channel (id=Num(4)), original request likely timed out
2024-06-16T13:39:31.975 helix_lsp::transport [ERROR] Tried sending response into a closed channel (id=Num(14)), original request likely timed out
2024-06-16T13:39:31.975 helix_lsp::transport [ERROR] Tried sending response into a closed channel (id=Num(12)), original request likely timed out
2024-06-16T13:39:32.134 helix_lsp::transport [ERROR] Tried sending response into a closed channel (id=Num(10)), original request likely timed out
2024-06-16T13:39:32.134 helix_lsp::transport [ERROR] Tried sending response into a closed channel (id=Num(7)), original request likely timed out
2024-06-16T13:39:32.135 helix_lsp::transport [ERROR] Tried sending response into a closed channel (id=Num(11)), original request likely timed out
2024-06-16T13:39:32.135 helix_lsp::transport [ERROR] Tried sending response into a closed channel (id=Num(13)), original request likely timed out
2024-06-16T13:42:45.863 helix_lsp::transport [ERROR] Tried sending response into a closed channel (id=Num(48)), original request likely timed out
2024-06-16T13:44:46.701 helix_lsp::transport [ERROR] Tried sending response into a closed channel (id=Num(58)), original request likely timed out
2024-06-16T13:44:47.356 helix_lsp::transport [ERROR] Tried sending response into a closed channel (id=Num(60)), original request likely timed out
2024-06-16T13:46:14.774 helix_lsp::transport [ERROR] Tried sending response into a closed channel (id=Num(76)), original request likely timed out
2024-06-16T13:46:44.090 helix_lsp::transport [ERROR] Tried sending response into a closed channel (id=Num(112)), original request likely timed out
2024-06-16T13:46:44.092 helix_lsp::transport [ERROR] Tried sending response into a closed channel (id=Num(110)), original request likely timed out
2024-06-16T13:47:30.688 helix_lsp::transport [ERROR] Tried sending response into a closed channel (id=Num(151)), original request likely timed out
2024-06-16T13:47:56.506 helix_lsp::transport [ERROR] Tried sending response into a closed channel (id=Num(165)), original request likely timed out
2024-06-16T13:49:58.652 helix_lsp::transport [ERROR] Tried sending response into a closed channel (id=Num(194)), original request likely timed out
2024-06-16T13:49:58.688 helix_lsp::transport [ERROR] Tried sending response into a closed channel (id=Num(193)), original request likely timed out
2024-06-16T13:50:55.797 helix_lsp::transport [ERROR] Tried sending response into a closed channel (id=Num(230)), original request likely timed out
2024-06-16T13:51:36.379 helix_lsp::transport [ERROR] Tried sending response into a closed channel (id=Num(242)), original request likely timed out
2024-06-16T13:51:39.991 helix_lsp::transport [ERROR] Tried sending response into a closed channel (id=Num(245)), original request likely timed out
2024-06-16T13:51:39.992 helix_lsp::transport [ERROR] Tried sending response into a closed channel (id=Num(246)), original request likely timed out
2024-06-16T13:51:40.023 helix_lsp::transport [ERROR] Tried sending response into a closed channel (id=Num(244)), original request likely timed out
2024-06-16T13:51:42.954 helix_lsp::transport [ERROR] Tried sending response into a closed channel (id=Num(254)), original request likely timed out
2024-06-16T13:51:59.322 helix_lsp::transport [ERROR] Tried sending response into a closed channel (id=Num(268)), original request likely timed out
2024-06-16T13:52:34.919 helix_lsp::transport [ERROR] Tried sending response into a closed channel (id=Num(285)), original request likely timed out
2024-06-16T13:52:34.920 helix_lsp::transport [ERROR] Tried sending response into a closed channel (id=Num(286)), original request likely timed out
2024-06-16T13:52:34.923 helix_lsp::transport [ERROR] Tried sending response into a closed channel (id=Num(283)), original request likely timed out
2024-06-16T13:52:34.925 helix_lsp::transport [ERROR] Tried sending response into a closed channel (id=Num(284)), original request likely timed out
2024-06-16T13:52:34.965 helix_lsp::transport [ERROR] Tried sending response into a closed channel (id=Num(282)), original request likely timed out
2024-06-16T13:52:37.684 helix_lsp::transport [ERROR] rust-analyzer err <- "2024-06-16T11:52:37.684587Z ERROR rust_analyzer::main_loop: flycheck 0: File with cargo diagnostic not found in VFS: file not found: /rustc/804421dff5542c9c7da5c60257b5dbc849719505/library/core/src/iter/traits/iterator.rs\n"
2024-06-16T13:52:37.684 helix_lsp::transport [ERROR] rust-analyzer err <- "2024-06-16T11:52:37.684954Z ERROR rust_analyzer::main_loop: flycheck 0: File with cargo diagnostic not found in VFS: file not found: /rustc/804421dff5542c9c7da5c60257b5dbc849719505/library/core/src/iter/traits/iterator.rs\n"
2024-06-16T13:52:44.755 helix_lsp::transport [ERROR] rust-analyzer err <- "2024-06-16T11:52:44.755039Z ERROR rust_analyzer::main_loop: flycheck 0: File with cargo diagnostic not found in VFS: file not found: /rustc/804421dff5542c9c7da5c60257b5dbc849719505/library/core/src/iter/traits/iterator.rs\n"
2024-06-16T13:52:44.755 helix_lsp::transport [ERROR] rust-analyzer err <- "2024-06-16T11:52:44.755157Z ERROR rust_analyzer::main_loop: flycheck 0: File with cargo diagnostic not found in VFS: file not found: /rustc/804421dff5542c9c7da5c60257b5dbc849719505/library/core/src/iter/traits/iterator.rs\n"
2024-06-16T13:54:40.025 helix_lsp::transport [ERROR] Tried sending response into a closed channel (id=Num(312)), original request likely timed out
2024-06-16T14:00:04.590 helix_lsp::transport [ERROR] rust-analyzer err <- "2024-06-16T12:00:04.590733Z ERROR hir::diagnostics: synthetic syntax\n"
2024-06-16T14:00:04.788 helix_lsp::transport [ERROR] rust-analyzer err <- "2024-06-16T12:00:04.788418Z ERROR hir::diagnostics: synthetic syntax\n"
2024-06-16T14:00:04.856 helix_lsp::transport [ERROR] rust-analyzer err <- "2024-06-16T12:00:04.856036Z ERROR hir::diagnostics: synthetic syntax\n"
2024-06-16T14:00:05.070 helix_lsp::transport [ERROR] rust-analyzer err <- "2024-06-16T12:00:05.070707Z ERROR hir::diagnostics: synthetic syntax\n"
2024-06-16T14:00:05.139 helix_lsp::transport [ERROR] rust-analyzer err <- "2024-06-16T12:00:05.139305Z ERROR hir::diagnostics: synthetic syntax\n"
2024-06-16T14:00:05.477 helix_lsp::transport [ERROR] rust-analyzer err <- "2024-06-16T12:00:05.477033Z ERROR hir::diagnostics: synthetic syntax\n"
2024-06-16T14:00:05.631 helix_lsp::transport [ERROR] rust-analyzer err <- "2024-06-16T12:00:05.631272Z ERROR hir::diagnostics: synthetic syntax\n"
2024-06-16T14:00:05.858 helix_lsp::transport [ERROR] rust-analyzer err <- "2024-06-16T12:00:05.858617Z ERROR hir::diagnostics: synthetic syntax\n"
2024-06-16T14:00:05.996 helix_lsp::transport [ERROR] rust-analyzer err <- "2024-06-16T12:00:05.996908Z ERROR hir::diagnostics: synthetic syntax\n"
2024-06-16T14:00:08.134 helix_lsp::transport [ERROR] rust-analyzer err <- "2024-06-16T12:00:08.134073Z ERROR hir::diagnostics: synthetic syntax\n"
2024-06-16T14:00:08.304 helix_lsp::transport [ERROR] rust-analyzer err <- "2024-06-16T12:00:08.304877Z ERROR hir::diagnostics: synthetic syntax\n"
2024-06-16T14:00:08.504 helix_lsp::transport [ERROR] rust-analyzer err <- "2024-06-16T12:00:08.504724Z ERROR hir::diagnostics: synthetic syntax\n"
2024-06-16T14:00:08.579 helix_lsp::transport [ERROR] rust-analyzer err <- "2024-06-16T12:00:08.579901Z ERROR hir::diagnostics: synthetic syntax\n"
2024-06-16T14:00:09.008 helix_lsp::transport [ERROR] rust-analyzer err <- "2024-06-16T12:00:09.008948Z ERROR hir::diagnostics: synthetic syntax\n"
2024-06-16T14:00:09.154 helix_lsp::transport [ERROR] rust-analyzer err <- "2024-06-16T12:00:09.153969Z ERROR hir::diagnostics: synthetic syntax\n"
2024-06-16T14:00:09.379 helix_lsp::transport [ERROR] rust-analyzer err <- "2024-06-16T12:00:09.379878Z ERROR hir::diagnostics: synthetic syntax\n"
2024-06-16T14:00:09.482 helix_lsp::transport [ERROR] rust-analyzer err <- "2024-06-16T12:00:09.482015Z ERROR hir::diagnostics: synthetic syntax\n"
```

</details>


### Platform

Linux (NixOS e9ee548d90ff586a6471b4ae80ae9cfcbceb3420)

### Terminal Emulator

alacritty 0.13.2

### Installation Method

flake

### Helix Version

helix 24.3 (dbacaaddcaa4a9dd39c73d0102d03ea7c24ca647)

--- a/helix-core/src/syntax.rs
+++ b/helix-core/src/syntax.rs
@@ -1363,13 +1363,14 @@ impl Syntax {
                 let depth = layer.depth + 1;
                 // TODO: can't inline this since matches borrows self.layers
                 for (config, ranges) in injections {
+                    let parent = Some(layer_id);
                     let new_layer = LanguageLayer {
                         tree: None,
                         config,
                         depth,
                         ranges,
                         flags: LayerUpdateFlags::empty(),
-                        parent: Some(layer_id),
+                        parent: None,
                     };
 
                     // Find an identical existing layer
@@ -1381,6 +1382,7 @@ impl Syntax {
 
                     // ...or insert a new one.
                     let layer_id = layer.unwrap_or_else(|| self.layers.insert(new_layer));
+                    self.layers[layer_id].parent = parent;
 
                     queue.push_back(layer_id);
                 }

